ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/ContractDao.scala

```diff
@@ -30,36 +33,93 @@ import scalaz.syntax.order._
 import spray.json.{JsNull, JsValue}
 
 import java.io.{Closeable, IOException}
+import java.sql.SQLTransientConnectionException
 import java.util.concurrent.{ExecutorService, Executors, TimeUnit}
 import javax.sql.DataSource
 import scala.concurrent.ExecutionContext
 import scala.language.existentials
-import scala.util.Try
+import scala.util.{Failure, Success, Try}
+import scala.annotation.nowarn
 
 class ContractDao private (
     ds: DataSource with Closeable,
     xa: ConnectionPool.T,
     dbAccessPool: ExecutorService,
+    @nowarn diagnosticSetup: Option[
+      (DataSource with Closeable, ConnectionPool.T, ExecutorService)
+    ] = None,
+    @nowarn diagnosticConfig: Option[DiagnosticConfig] = None,
 )(implicit
     val jdbcDriver: SupportedJdbcDriver.TC
 ) extends Closeable {
 
   private val logger = LoggerFactory.getLogger(classOf[ContractDao])
   implicit val logHandler: log.LogHandler = Slf4jLogHandler(logger)
 
-  def transact[A](query: ConnectionIO[A]): IO[A] =
-    query.transact(xa)
+  def transact[A](query: ConnectionIO[A], clientInfo: Option[String] = None): IO[A] = {
+    val xa0 = (diagnosticConfig.flatMap(_.clientInfo), clientInfo) match {
+      case (Some(identifier), Some(info)) =>
+        Transactor.strategy.modify(
+          xa,
+          s => s.copy(before = s.before.flatMap(_ => setClientInfo(identifier, info))),
+        )
+      case _ => xa
+    }
+    diagnosticSetup match {
+      case Some((_, dxa, _)) =>
+        query.transact(xa0).handleErrorWith {
+          case e: SQLTransientConnectionException
+              if e.getMessage.contains("Connection is not available") =>
+            runDiagnostic(diagnosticConfig, dxa)
+            IO.raiseError(e)
+          case e => IO.raiseError(e)
+        }
+      case _ => query.transact(xa0)
+    }
+  }
+
+  private def runDiagnostic(diagnosticConfig: Option[DiagnosticConfig], dxa: ConnectionPool.T) = {
```

Comment:
```suggestion
  private def runDiagnostic(diagnosticConfig: DiagnosticConfig, dxa: ConnectionPool.T) = {
```
I think better to do the `foreach` a level up, and then this function is more straight-forward.